intrinsics fmuladdf{32,64}: expose llvm.fmuladd.* semantics

jedbrown opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/124874)*
<!-- homu-ignore:end -->

tracking issue: https://github.com/rust-lang/rust/issues/151770

Add intrinsics `fmuladd{f32,f64}`. This computes `(a * b) + c`, to be fused if the code generator determines that (i) the target instruction set has support for a fused operation, and (ii) that the fused operation is more efficient than the equivalent, separate pair of `mul` and `add` instructions.

https://llvm.org/docs/LangRef.html#llvm-fmuladd-intrinsic

The codegen_cranelift uses the `fma` function from libc, which is a correct implementation, but without the desired performance semantic. I think this requires an update to cranelift to expose a suitable instruction in its IR.

I have not tested with codegen_gcc, but it should behave the same way (using `fma` from libc).

---
This topic has been discussed a few times on Zulip and was suggested, for example, by @workingjubilee in [Effect of fma disabled](https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/Effect.20of.20fma.20disabled/near/274179331).

jedbrown committed:
intrinsics.fmuladdf{16,32,64,128}: expose llvm.fmuladd.* semantics

Add intrinsics `fmuladd{f16,f32,f64,f128}`. This computes `(a * b) +
c`, to be fused if the code generator determines that (i) the target
instruction set has support for a fused operation, and (ii) that the
fused operation is more efficient than the equivalent, separate pair
of `mul` and `add` instructions.

https://llvm.org/docs/LangRef.html#llvm-fmuladd-intrinsic

MIRI support is included for f32 and f64.

The codegen_cranelift uses the `fma` function from libc, which is a
correct implementation, but without the desired performance semantic. I
think this requires an update to cranelift to expose a suitable
instruction in its IR.

I have not tested with codegen_gcc, but it should behave the same
way (using `fma` from libc).

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8995617868/job/24710871709) [(plain)](https://github.com/rust-lang/rust/commit/5de9a1e5a4676976d7e371fac2c95f4e505565cc/checks/24710871709/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

saethlin commented on library/std/src/f32.rs:
```diff
@@ -293,6 +293,34 @@ impl f32 {
+    #[cfg_attr(not(bootstrap), rustc_allow_incoherent_impl)]
+    #[must_use = "method returns a new number and does not mutate the original value"]
+    #[inline]
+    #[unstable(feature = "float_mul_add_fast", issue = "none")]
+    pub fn mul_add_fast(self, a: f32, b: f32) -> f32 {
```
I'd rather we not call so many things "fast". Standard fast-math has so much UB that it's nearly unusable, so at the request of some people in the community I split out separate intrinsics that offer optimizable semanatics but are safe: https://github.com/rust-lang/rust/pull/120718

Do you have examples of this intrinsic offering better optimizations than `fadd_algebraic` and `fmul_algebraic`?

workingjubilee replied:
it could simply be `mul_add2`, tbh, given it is exactly multiply-add but allows 2 roundings.

workingjubilee replied:
it is somewhat undesirable that we already used `mul_add` to indicate the 1-rounding variant because it is common to use `fma` as the fused (1-rounding) version and `mul_add` as the 1-or-2 rounding version.

RalfJung replied:
> because it is common to use fma as the fused (1-rounding) version and mul_add as the 1-or-2 rounding version.

Ah that explains the terrible names of the new intrinsics, I guess.^^ It is not self-explaining at all...

jedbrown replied:
@saethlin I think your `_algebraic` intrinsics are a better fit for type-based techniques like [fast-floats](https://github.com/bluss/fast-floats), but it interacts differently with its surrounding. For example, consider $a(b/a) + a$
```rs
#[inline(never)]
fn foo_mul_add_fast(a: f64, b: f64) -> f64 {
    a.mul_add_fast(b / a, a)
}

#[inline(never)]
fn foo_algebraic(a: f64, b: f64) -> f64 {
    use std::intrinsics::fadd_algebraic as fadd;
    use std::intrinsics::fmul_algebraic as fmul;
    fadd(fmul(a, b / a), a)
}
```
Then the first does what is I think is intuitively expected while the second reformulates to $(ab)/a + a$ without any fmadd.
```console
$ RUSTFLAGS='-Ctarget-cpu=skylake' cargo +stage1 asm foo_mul_add_fast
[...]
        vdivsd xmm1, xmm1, xmm0
        vfmadd231sd xmm0, xmm1, xmm0
$ RUSTFLAGS='-Ctarget-cpu=skylake' cargo +stage1 asm foo_algebraic
[...]
        vmulsd xmm1, xmm1, xmm0
        vdivsd xmm1, xmm1, xmm0
        vaddsd xmm0, xmm1, xmm0
```
These are meaningfully different in case of overflow/underflow. Consider calling `foo_*(x, x)`:
* if `x = 1e300_f64`, the former is exact while the latter is `inf`
* if `x = 1e-300_f64`, the former is exact while the latter is off by a factor of two.

Regarding name, I'm happy with whatever has consensus.

saethlin replied:
That is interesting. Those intrinsics set the LLVM fast-math flags `reassoc` and `contract`, and as someone who never really studied floating-point details, it seems like those authorize creating the fused multiply-add. Is this a missing optimization in LLVM, or do I misunderstand?

jedbrown replied:
Yeah, this example produces identical code if we remove the division, so the contraction really does happen (but not at `opt-level=0` for `*_algebraic`, while it does use `vfmadd*` at `opt-level=0` when `llvm.fmuladd.*` is used). I don't know whether this transformation for `foo_algebraic` is justified for the microarchitecture (e.g., which execution unit will be tied up while OOO executing subsequent instructions) or not.

While initial lowering of `llvm.fmuladd` is keyed on `isFMAFasterThanFMulAndFAdd` (which depends on the target and FP mode), const-folding makes it dependent on optimization level. From [Limit FMul constant folding for fma simplifications](https://reviews.llvm.org/D67434):
> My understanding is that as long as it is a fmuladd, the code generator did not make a decision about fusing yet. IIUC there are 2 ways it can decide to fuse: 1) replace it with an fma intrinsic or lower it to a FMA in the backend.

That review spun off the [clarifying language](https://reviews.llvm.org/D67552) and [introduced the `@fmuladd_const_fmul` test](https://github.com/llvm/llvm-project/commit/f3ab99dcf8afa#diff-0f2df0f2960d65283247cddf3e7ffe319b36da72957df15c2eedbffdab01e6f9R503-R512) lowering `fmuladd` to `fadd` when the multiply part ($a, b$ in $a b + c$) is const.

Empirically, it is useful that`llvm.fmuladd` typically produces the same result in debug and optimized builds, but I can provoke the optimization dependence using const terms.
```rs
fn t_mul_add_fast(c: f64) -> f64 {
    let a = 0.1_f64;
    let b = 0.1_f64;
    a.mul_add_fast(b, c)
}
```
and calling as
```rs
    let x = std::hint::black_box(0.1_f64);
    let y = t_mul_add_fast(-x * x);
```
With `RUSTFLAGS='-Ctarget-cpu=skylake -Copt-level=0'`, this uses an FMA and produces `-8.326672684688674e-19`, but at `opt-level=1`, the const part is computed at compile-time, leaving only `vaddsd`, and thus evaluating to `0.0`. I think this removes any doubt about nondeterminism.

jedbrown replied:
I'll resolve this thread since I think it's now off-topic given that I've reduced scope to only adding the intrinsics. I think there are some design choices here if one wants to have `_algebraic` variants and use of `fence`; perhaps make a new issue or move discussion to Zulip?

jhpratt commented:
Given that this is dealing primarily with intrinsics, I think this is T-compiler land? In any situation this isn't my area of knowledge.

r? compiler

workingjubilee commented:
T-compiler? hmm, in implementation, but

hey @rust-lang/opsem we're discussing adding nondeterministic float rounding.

workingjubilee commented on library/std/src/f32/tests.rs:
```diff
@@ -409,6 +409,22 @@ fn test_mul_add() {
     assert_eq!((-3.2f32).mul_add(2.4, neg_inf), neg_inf);
 }
 
+#[test]
+fn test_mul_add_fast() {
+    let nan: f32 = f32::NAN;
+    let inf: f32 = f32::INFINITY;
+    let neg_inf: f32 = f32::NEG_INFINITY;
+    assert_approx_eq!(12.3f32.mul_add_fast(4.5, 6.7), 62.05);
+    assert_approx_eq!((-12.3f32).mul_add_fast(-4.5, -6.7), 48.65);
+    assert_approx_eq!(0.0f32.mul_add_fast(8.9, 1.2), 1.2);
+    assert_approx_eq!(3.4f32.mul_add_fast(-0.0, 5.6), 5.6);
+    assert!(nan.mul_add_fast(7.8, 9.0).is_nan());
+    assert_eq!(inf.mul_add_fast(7.8, 9.0), inf);
+    assert_eq!(neg_inf.mul_add_fast(7.8, 9.0), neg_inf);
+    assert_eq!(8.9f32.mul_add_fast(inf, 3.2), inf);
+    assert_eq!((-3.2f32).mul_add_fast(2.4, neg_inf), neg_inf);
+}
```
I believe this test is being compiled with the bootstrap stdlib, but these names don't exist in the bootstrap stdlib yet, so it's failing.
```suggestion
#[cfg(not(bootstrap))]
#[test]
fn test_mul_add_fast() {
    let nan: f32 = f32::NAN;
    let inf: f32 = f32::INFINITY;
    let neg_inf: f32 = f32::NEG_INFINITY;
    assert_approx_eq!(12.3f32.mul_add_fast(4.5, 6.7), 62.05);
    assert_approx_eq!((-12.3f32).mul_add_fast(-4.5, -6.7), 48.65);
    assert_approx_eq!(0.0f32.mul_add_fast(8.9, 1.2), 1.2);
    assert_approx_eq!(3.4f32.mul_add_fast(-0.0, 5.6), 5.6);
    assert!(nan.mul_add_fast(7.8, 9.0).is_nan());
    assert_eq!(inf.mul_add_fast(7.8, 9.0), inf);
    assert_eq!(neg_inf.mul_add_fast(7.8, 9.0), neg_inf);
    assert_eq!(8.9f32.mul_add_fast(inf, 3.2), inf);
    assert_eq!((-3.2f32).mul_add_fast(2.4, neg_inf), neg_inf);
}
```

workingjubilee commented on library/std/src/f64/tests.rs:
```diff
@@ -397,6 +397,22 @@ fn test_mul_add() {
     assert_eq!((-3.2f64).mul_add(2.4, neg_inf), neg_inf);
 }
 
+#[test]
+fn test_mul_add_fast() {
+    let nan: f64 = f64::NAN;
+    let inf: f64 = f64::INFINITY;
+    let neg_inf: f64 = f64::NEG_INFINITY;
+    assert_approx_eq!(12.3f64.mul_add_fast(4.5, 6.7), 62.05);
+    assert_approx_eq!((-12.3f64).mul_add_fast(-4.5, -6.7), 48.65);
+    assert_approx_eq!(0.0f64.mul_add_fast(8.9, 1.2), 1.2);
+    assert_approx_eq!(3.4f64.mul_add_fast(-0.0, 5.6), 5.6);
+    assert!(nan.mul_add_fast(7.8, 9.0).is_nan());
+    assert_eq!(inf.mul_add_fast(7.8, 9.0), inf);
+    assert_eq!(neg_inf.mul_add_fast(7.8, 9.0), neg_inf);
+    assert_eq!(8.9f64.mul_add_fast(inf, 3.2), inf);
+    assert_eq!((-3.2f64).mul_add_fast(2.4, neg_inf), neg_inf);
+}
```
```suggestion
#[cfg(not(bootstrap))]
#[test]
fn test_mul_add_fast() {
    let nan: f64 = f64::NAN;
    let inf: f64 = f64::INFINITY;
    let neg_inf: f64 = f64::NEG_INFINITY;
    assert_approx_eq!(12.3f64.mul_add_fast(4.5, 6.7), 62.05);
    assert_approx_eq!((-12.3f64).mul_add_fast(-4.5, -6.7), 48.65);
    assert_approx_eq!(0.0f64.mul_add_fast(8.9, 1.2), 1.2);
    assert_approx_eq!(3.4f64.mul_add_fast(-0.0, 5.6), 5.6);
    assert!(nan.mul_add_fast(7.8, 9.0).is_nan());
    assert_eq!(inf.mul_add_fast(7.8, 9.0), inf);
    assert_eq!(neg_inf.mul_add_fast(7.8, 9.0), neg_inf);
    assert_eq!(8.9f64.mul_add_fast(inf, 3.2), inf);
    assert_eq!((-3.2f64).mul_add_fast(2.4, neg_inf), neg_inf);
}
```

workingjubilee commented on compiler/rustc_codegen_cranelift/src/intrinsics/mod.rs:
```diff
@@ -328,6 +328,8 @@ fn codegen_float_intrinsic_call<'tcx>(
         sym::fabsf64 => ("fabs", 1, fx.tcx.types.f64, types::F64),
         sym::fmaf32 => ("fmaf", 3, fx.tcx.types.f32, types::F32),
         sym::fmaf64 => ("fma", 3, fx.tcx.types.f64, types::F64),
+        sym::fmuladdf32 => ("fmaf", 3, fx.tcx.types.f32, types::F32), // NOTE: pessimal without FMA target feature
+        sym::fmuladdf64 => ("fma", 3, fx.tcx.types.f64, types::F64), // NOTE: pessimal without FMA target feature
```
"pessimal" is vague? it's not a pessimization in all cases.

jedbrown replied:
```suggestion
        sym::fmuladdf32 => ("fmaf", 3, fx.tcx.types.f32, types::F32), // NOTE: software emulation without FMA target feature
        sym::fmuladdf64 => ("fma", 3, fx.tcx.types.f64, types::F64), // NOTE: software emulation without FMA target feature
```

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8996263506/job/24712485445) [(plain)](https://github.com/rust-lang/rust/commit/a3ab7950f4057869c086842bc848535e9b1d2d68/checks/24712485445/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking ui_test v0.21.2
error[E0635]: unknown feature `float_mul_add_fast`
 --> src/tools/miri/src/lib.rs:5:12
  |
5 | #![feature(float_mul_add_fast)]

For more information about this error, try `rustc --explain E0635`.
error: could not compile `miri` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
```

</details>

RalfJung commented on library/core/src/intrinsics.rs:
```diff
@@ -1724,6 +1724,25 @@ extern "rust-intrinsic" {
+    /// and add instructions.
+    #[rustc_nounwind]
+    #[cfg(not(bootstrap))]
+    pub fn fmuladdf32(a: f32, b: f32, c: f32) -> f32;
```
So sounds like the spec is : non-deterministically either fuse or not? But the result is *guaranteed* to be either the fully precise fused computation rounded once, or the fully precise multiplication, rounded, then fully precise addition, rounded?

A spec should not usually mention "codegen". A spec should say what the behavior is. Codegen's job is to implement that.

When I call this multiple times with the same inputs in the same program, am I guaranteed to either always get the fused behavior or always the unfused behavior, or can it differ between different invocations? What if they used different inputs?

workingjubilee replied:
> So sounds like the spec is : non-deterministically either fuse or not? But the result is guaranteed to be either the fully precise fused computation rounded once, or the fully precise multiplication, rounded, then fully precise addition, rounded?

Correct.

> When I call this multiple times with the same inputs in the same program, am I guaranteed to either always get the fused behavior or always the unfused behavior, or can it differ between different invocations? What if they used different inputs?

That is an interesting question I don't know the answer to! :D

workingjubilee replied:
The spec in the [LangRef for fmuladd](https://llvm.org/docs/LangRef.html#id806) suggests full-throttle nondeterminism.

jedbrown replied:
I don't know a definitive answer, but I think it's unspecified (e.g., depending on instruction latency and the surrounding code, LLVM could choose separate versus fused instructions). I haven't been able to provoke that difference in ad-hoc experiments, but that's no surprise when FMA instructions have identical latency to pure multiplication.

RalfJung replied:
Yeah the LangRef seems to make it fully non-deterministic. So we should go with that for now, and in particular document this clearly in the public function.

RalfJung commented on src/tools/miri/src/lib.rs:
```diff
@@ -2,6 +2,7 @@
 #![feature(cell_update)]
 #![feature(const_option)]
 #![feature(float_gamma)]
+#![cfg_attr(not(bootstrap), feature(float_mul_add_fast))]
```
I don't think it is necessary to add this to Miri? It doesn't seem used.

jedbrown replied:
I removed the new feature from this PR (which now has only the intrinsics).

RalfJung commented on tests/ui/intrinsics/intrinsics-math.rs:
Please add a test to Miri as well.

RalfJung commented on library/core/src/intrinsics.rs:
```diff
@@ -1724,6 +1724,25 @@ extern "rust-intrinsic" {
+    /// and add instructions.
+    #[rustc_nounwind]
+    #[cfg(not(bootstrap))]
+    pub fn fmuladdf32(a: f32, b: f32, c: f32) -> f32;
```
Is this a good opportunity to use our fallback bodies, with a fallback implementation of `a* b + c`? Const-eval should also use that.

Miri however should probably roll some dice to make both behaviors observable.

jedbrown replied:
I don't have strong rationale for choosing the fused versus separate, but it was the simplest to implement and presumably to change if/when Cranelift adds a corresponding intrinsic.

Could you give a pointer for how/where to implement the "make both behaviors observable" in Miri?

RalfJung replied:
The miri implementation would go in `src/tools/miri/src/intrinsics/mod.rs`. There you can use `this.machine.rng.borrow_mut()` to get access to an RNG that you can use to decide whether to use the fused implementation or the double-rounding implementation. That way, if the function is called many times in Miri, both behaviors will be observable.

jedbrown replied:
I implemented this and added a test in the spirit of `is_val_statically_known` in `src/tools/miri/tests/pass/intrinsics/intrinsics.rs` (looping 50 times to observe both outcomes). Let me know if there is a better way to ensure that.

RalfJung commented:
Yeah full non-det is kind of what I would expect. The intrinsic can just say "nondeterministic" but the public function should probably spell this out in its docs.

antoyo commented on compiler/rustc_codegen_gcc/src/intrinsic/mod.rs:
```diff
@@ -66,6 +66,8 @@ fn get_simple_intrinsic<'gcc, 'tcx>(
         sym::log2f64 => "log2",
         sym::fmaf32 => "fmaf",
         sym::fmaf64 => "fma",
+        sym::fmuladdf32 => "fmaf", // NOTE: pessimal without FMA target feature
```
Can you please explain what do you mean by "pessimal without FMA target feature"?

jedbrown replied:
Thanks, I'll update accordingly. [This comment](https://github.com/rust-lang/rust/pull/124874#discussion_r1593464410) on whether to fall back to fused or separate.

workingjubilee commented:
@jedbrown So, T-libs-api gets to decide whether the name on f32/f64 is good, basically. It might be best to land the intrinsic and then PR the public safe API.

jedbrown commented:
> @jedbrown So, T-libs-api gets to decide whether the name on f32/f64 is good, basically. It might be best to land the intrinsic and then PR the public safe API.

Should I update this PR to only handle the intrinsic?

workingjubilee commented:
@jedbrown up to you but it would mean that only the people in charge of the compiler/opsem decide whether it's merged so a shorter path to getting this PR over and done with and then the bikeshedding, if any, can be on its own PR

clarfonthey commented:
Not to further delay merging, but given how this fits more along the line of the existing `wrapping` methods in the sense that it might be worth eventually adding a `Fast` wrapper type, perhaps `fast_mul_add` would be a better name?

Imagining there are many other operations that could get this treatment, like `hypot`.

saethlin commented:
The point of such types is to provide operator overloads that make using the type ergonomic. You can't do that with our `_fast` semantics because all the `_fast` operations are unsafe. You'd need an unsafe version of the operators. So existing fast float libraries just ignore this and are unsound or have uselessly broad preconditions such as https://docs.rs/fast-floats/0.2.0/fast_floats/struct.Fast.html#safety.

That's why I proposed the `_algebraic` intrinsics. Those make it possible to write a usable wrapper. You could of course RFC unsafe operator overloads to make the `_fast` version technically possible to write a sound API over, though I submit that in practice it's absurd to prove preconditions like "this float will never be NaN".

clarfonthey commented:
I guess that the issue is that "fast" is too overloaded with existing "fast float" semantics, although it really is the best term to use here. It's the only way to definitively summarise that, between fused multiplication and simple multiply-add, it chooses whatever is the fastest option, not necessarily the accurate one.

"Approximate" doesn't necessarily summarise this, since while it points out you're allowing approximations, it could imply it's worse than what you'd expect.

"Algebraic" honestly doesn't do it for me. To me, algebraic means specifically [algebraic numbers](https://en.wikipedia.org/wiki/Algebraic_number) and alludes to stuff like [taylor series](https://en.wikipedia.org/wiki/Taylor_series) being used to compute the result, which is not what you're doing here. I get that you mean "algebraic simplification" but the key word there is simplification, not algebraic.

On that end, "simple" might work as a compromise here, but I dunno. I really think that we could just use the term "fast" here to mean "accuracy may be sacrificed, but soundness won't be" and if we want to add in potentially-unsound versions, we could use the term "unsafe" or "unchecked" instead of fast.

CAD97 commented:
I just want to note that this naming "clashing" with `-ffast-math` style "fast" isn't fully *wrong* per say either, as the contraction from separate mul and add to fused mul/add is exactly the kind of optimization which the safe subset of fast-math flags are enabling. 

It seems very unfortunate to tie this one case to the more general bikeshed that is exposing fast math semantics more generally, but it is exactly the same naming question, just with the benefit of the safe subset (that doesn't assume that all is finite) much more clear than with each of the primitive ops separate.

Given the precedent of `unchecked_op` on the integer types meaning assume no wrapping, I can support `unchecked_op` meaning assume no non-finite for the floating point types. There are three categories of current fast-math opts:

- finite
  - assume no NaN (`nnan`; NaN becomes `poison`)
  - assume no Inf (`ninf`; ±Inf becomes `poison`)
- algebraic
  - no signed zeros (`nsz`; zero always has ndet sign)
  - allow reciprocal (`arcp`; minor ndet impact on precision)
  - allow contraction (`contract`; minor ndet impact on precision)
  - allow reassociation (`reassoc`; may ndet "dramatically" change results) 
- approx
  - allow function approximation (`afn`; may ndet significantly impact precision)

and "fast" means all of them. This function only wants contraction and only of the one mul/add operation; applying *any* more of these flags would not be correct as per the PR description.

Other potentially useful labels: weak, inexact, lenient, relaxed

-----

This restates a lot of the information but isn't all that helpful in resolving this specific naming. I'd suggest landing the intrinsics first and adding the public function veneer in a follow up PR as I don't personally expect this bikeshed to get its paint any time soon.

clarfonthey commented:
> I'd suggest landing the intrinsics first and adding the public function veneer in a follow up PR as I don't personally expect this bikeshed to get its paint any time soon.

Agreed! I hadn't even realised this was just the intrinsic and not the method version. I'm fine with this staying unstable under whatever name it gets as long as it's documented.

RalfJung commented on library/core/src/intrinsics.rs:
```diff
@@ -1803,6 +1803,29 @@ extern "rust-intrinsic" {
     #[rustc_nounwind]
     pub fn fmaf128(a: f128, b: f128, c: f128) -> f128;
 
+    /// Returns `a * b + c` for `f32` values.
```
```suggestion
    /// Returns `a * b + c` for `f32` values, non-deterministically executing either a
    /// fused multiply-add or two operations with rounding of the intermediate result.
```
Same for the other one.

jedbrown replied:
Done.

RalfJung commented on src/tools/miri/src/intrinsics/mod.rs:
```diff
@@ -308,6 +308,39 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+                    let res = a.to_host().mul_add(b.to_host(), c.to_host()).to_soft();
+                    this.adjust_nan(res, &[a, b, c])
+                } else {
+                    let res = (a.to_host() * b.to_host() + c.to_host()).to_soft();
```
You can use portable floats for this arm. (And then move the FIXME into the `fuse` arm.)

jedbrown replied:
Thanks, I think I've done what you intended.

RalfJung commented on src/tools/miri/src/intrinsics/mod.rs:
```diff
@@ -308,6 +308,41 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+                let res = if fuse {
+                    // FIXME: Using host floats, to work around https://github.com/rust-lang/rustc_apfloat/issues/11
+                    let res = a.to_host().mul_add(b.to_host(), c.to_host()).to_soft();
+                    this.adjust_nan(res, &[a, b, c])
```
The `adjust_nan` can be moved out of the `if`

RalfJung commented on src/tools/miri/src/intrinsics/mod.rs:
```diff
@@ -308,6 +308,41 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+                    this.adjust_nan(res, &[a, b, c])
+                } else {
+                    let round = Round::NearestTiesToEven;
+                    let res = a.mul_r(b, round).value.add_r(c, round).value;
```
`((a*b).value + c).value` should also work. Or `a.mul(b).value.add(c).value`, if you prefer that. But no reason to set the rounding mode explicitly.

jedbrown replied:
Oh, yes. Much nicer.

jedbrown commented:
Sorry about the big delay getting back to this. I've updated this PR to remove the `float_mul_add_fast` feature and associated safe functions, keeping the `intrinsics::fmuladdf{32,64}` with associated tests and Miri support (thanks Ralf, for the guidance). I think it's reviewable now.

jedbrown commented:
Hi :wave: Just checking in about when this can be reviewed. I'm unclear if it needs all the listed reviewers or just a subset since it's been pared down.

RalfJung commented:
There's a single assigned reviewer. We can roll a new one.

r? compiler

nnethercote commented:
lol, I don't think randomly rolling for a review on a PR that is 4.5 months old and has dozens of comments from people with r+ permissions is a good idea. I know very little about floating point arithmetic, so I will pass this on to one of the people who does know about that topic and has already weighed in:

r? @saethlin

saethlin commented:
Regardless of answer to that question, I think this is good to land. These intrinsics should be useful in the community for building out libraries that hopefully can drive proper Rust support for fast floats.

@bors r+

bors commented:
:pushpin: Commit c0744ff655fe01a637330069ff06caa9bf39caa1 has been approved by `saethlin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=saethlin c0744ff655fe01a637330069ff06caa9bf39caa1 -->
<!-- homu: {"type":"Approved","sha":"c0744ff655fe01a637330069ff06caa9bf39caa1","approver":"saethlin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

RalfJung commented:
Ah yeah seems like mostly the f16 / f128 intrinsics are declared in libcore, so we should declare these as well to make sure the docs are in sync with the codegen backend. But that can be a follow-up PR.

bors commented:
:hourglass: Testing commit c0744ff655fe01a637330069ff06caa9bf39caa1 with merge 8f1fa6788194ded866e4b351a5828e7ea638c3e2...
<!-- homu: {"type":"BuildStarted","head_sha":"c0744ff655fe01a637330069ff06caa9bf39caa1","merge_sha":"8f1fa6788194ded866e4b351a5828e7ea638c3e2"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-nopt`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10958106141/job/30427667858) [(plain)](https://github.com/rust-lang-ci/rust/commit/8f1fa6788194ded866e4b351a5828e7ea638c3e2/checks/30427667858/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [ui] tests/ui/intrinsics/intrinsic-fmuladd.rs stdout ----

error: test run failed!
status: exit status: 101
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/intrinsics/intrinsic-fmuladd" && RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUST_TEST_THREADS="4" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/intrinsics/intrinsic-fmuladd/a"
--- stderr -------------------------------
thread 'main' panicked at /checkout/tests/ui/intrinsics/intrinsic-fmuladd.rs:18:9:
62.050003 is not approximately equal to 62.05
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10958106141/job/30429634360)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/10958106141/job/30429634360","builder_name":"checks-actions"} -->

jedbrown commented:
I rebased and updated this PR to include the `f16` and `f128` intrinsics (thanks @saethlin), and I can use them when I enable the corresponding features, with or without hardware support (e.g., `vfmadd213sh` are issued for `f16` with `target-cpu=sapphirerapids`).

I also reduced the arguments for the test by one digit because it was just tripping f32 machine epsilon so I expect CI to pass this time.

saethlin commented:
@bors r+

bors commented:
:pushpin: Commit 0d8a978e8a55b08778ec6ee861c2c5ed6703eb6c has been approved by `saethlin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=saethlin 0d8a978e8a55b08778ec6ee861c2c5ed6703eb6c -->
<!-- homu: {"type":"Approved","sha":"0d8a978e8a55b08778ec6ee861c2c5ed6703eb6c","approver":"saethlin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
